### PR TITLE
feat: fix update public plan proposal & disable codecov patch

### PR DESCRIPTION
**Pull request by kogisin**
## Description

closes: #131 

## Tasks

- [x] Change [UpdatePublicPlanProposal](https://github.com/tendermint/farming/blob/master/x/farming/keeper/proposal_handler.go#L99) logic to update plan type when there is a request
- [x] Add test codes to cover the updated cases
- [x] disable codecov patch (reference the comments below)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

**Comment by kogisin** [comment]
codecov patch is not reliable and there are many issues out there proposing it to be disabled.
- https://github.com/kata-containers/runtime/issues/511
- https://github.com/kata-containers/tests/issues/508
https://github.com/aws/amazon-vpc-cni-k8s/pull/1226/files

File: x/farming/keeper/proposal_handler.go [hunk 1]
```diff
@@ -103,8 +105,13 @@ func (k Keeper) UpdatePublicPlanProposal(ctx sdk.Context, proposals []*types.Upd
 			return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "plan %d is not found", p.GetPlanId())
 		}
 
-		switch plan := plan.(type) {
-		case *types.FixedAmountPlan:
+		if p.EpochAmount.IsAllPositive() {
+			if p.GetName() != "" {
+				if err := plan.SetName(p.GetName()); err != nil {
```
Comment (dongsam) [inline x/farming/keeper/proposal_handler.go:110]: currently `SetName` always return `nil`, so how about using `plan.Name = p.GetName()` for the coverage?

File: x/farming/keeper/proposal_handler.go [hunk 2]
```diff
@@ -143,22 +150,32 @@ func (k Keeper) UpdatePublicPlanProposal(ctx sdk.Context, proposals []*types.Upd
...
 
 			logger := k.Logger(ctx)
 			logger.Info("updated public fixed amount plan", "fixed_amount_plan", plan)
 
-		case *types.RatioPlan:
-			if err := plan.Validate(); err != nil {
-				return err
+		} else if p.EpochRatio.IsPositive() {
+			if p.GetName() != "" {
+				if err := plan.SetName(p.GetName()); err != nil {
```
Comment (dongsam) [inline x/farming/keeper/proposal_handler.go:176]: same above

File: x/farming/keeper/proposal_handler.go (hunk 1 above)
Comment (kogisin) [reply x/farming/keeper/proposal_handler.go:110]: As discussed offline, I'll make the changes if I can't find a solution to disable codecov/patch.

**Review by dongsam — APPROVED** [review]
LGTM!

File: x/farming/keeper/proposal_handler.go [hunk 3]
```diff
@@ -103,8 +105,13 @@ func (k Keeper) UpdatePublicPlanProposal(ctx sdk.Context, proposals []*types.Upd
 			return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "plan %d is not found", p.GetPlanId())
 		}
 
-		switch plan := plan.(type) {
-		case *types.FixedAmountPlan:
+		if p.EpochAmount.IsAllPositive() {
```
Comment (hallazzang) [inline x/farming/keeper/proposal_handler.go:108]: We should check if both `EpochAmount` and `EpochRatio` are set at the same time. If both of those fields were set, then it should reject the proposal to disambiguate.

File: x/farming/keeper/proposal_handler.go [hunk 4]
```diff
@@ -143,22 +150,32 @@ func (k Keeper) UpdatePublicPlanProposal(ctx sdk.Context, proposals []*types.Upd
...
 			}
 
-			if p.GetName() != "" {
-				plan.Name = p.GetName()
-			}
-
-			if p.GetEpochAmount() != nil {
-				plan.EpochAmount = p.GetEpochAmount()
+			// change the plan to fixed amount plan if an epoch amount exists
+			if p.GetEpochAmount().IsAllPositive() {
```
Comment (hallazzang) [inline x/farming/keeper/proposal_handler.go:154]: It's not necessary but for simplicity, we could add `GetBasePlan` method to the `PlanI` interface and use that here as:
```go
plan = types.NewFixedAmountPlan(plan.GetBasePlan(), p.GetEpochAmount())
```

**Review by hallazzang — CHANGES_REQUESTED** [review]

File: x/farming/keeper/proposal_handler.go (hunk 3 above)
Comment (kogisin) [reply x/farming/keeper/proposal_handler.go:108]: I think that is being handled in [ValidateBasic](https://github.com/tendermint/farming/blob/master/x/farming/types/proposal.go#L47) when a user sends a governance proposal, isn't it?

File: x/farming/keeper/proposal_handler.go (hunk 4 above)
Comment (kogisin) [reply x/farming/keeper/proposal_handler.go:154]: Yes, I agree that makes it simpler. Let me make that change.

File: x/farming/keeper/proposal_handler.go (hunk 3 above)
Comment (hallazzang) [reply x/farming/keeper/proposal_handler.go:108]: Oh, ok. You're correct.

**Review by hallazzang — APPROVED** [review]

**Pull request merged**